refactor: BlockList uses Hooks

dcalhoun opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Refactor the mobile `BlockList` component to use Hooks.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Relates to https://github.com/wordpress-mobile/gutenberg-mobile/issues/5617. 

Adopt modern React paradigms prior to refactoring the components
composition. Completed to more closely mirror other code in the code
base.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
Replace `BlockList` and `EmptyList` class components with React Hooks. Remove an apparent unused inclusion of `withPreferredColorScheme`. 

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
Given this changes foundational code for the block list, we should likely test
adding/editing/relocating/removing blocks in various ways, including VoiceOver
and TalkBack. We should also be mindful of any performance regressions. 

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
n/a

## Screenshots or screencast <!-- if applicable -->
n/a

dcalhoun committed:
refactor: BlockList EmptyList uses Hooks

Adopt modern React paradigms prior to refactoring the components
composition. Completed to more closely mirror other code in the code
base.

dcalhoun committed:
refactor: BlockList uses Hooks

Adopt modern React paradigms prior to refactoring the components
composition. Completed to more closely mirror other code in the code
base.

dcalhoun committed:
refactor: BlockList higher-order components use Hooks

Adopt modern React paradigms prior to refactoring the components
composition. Completed to more closely mirror other code in the code
base.

Additionally, this removes the `withPreferredColorScheme` higher-order
component as it appeared unused.

dcalhoun committed:
fix: BlockList footer interactivity

`VirtualList` expects a React Component (e.g. `SomeComponent`), or a
React element (e.g. `<SomeComponent` />) passed to the
`ListFooterComponent` prop. Passing a function instead resulted in the
block inserter never opening when tapping the block appender button
within the footer; this appeared to be due to excessive re-renders
resetting the `isOpen` state for the inserter.

dcalhoun committed:
fix: Rename Footer addBlockToEndOfPost callback prop

The typo led to an undefined callback function.

dcalhoun committed:
fix: Reinstate missing prop used within store selector

This prop was originally accessed via the `withSelect` higher-order
component. When converting to Hooks, passing this prop was overlooked.

dcalhoun commented on packages/block-editor/src/components/block-list/block-list-item-cell.native.js:
```diff
@@ -7,14 +7,23 @@ import { View } from 'react-native';
-function BlockListItemCell( { children, clientId, rootClientId, onLayout } ) {
+function BlockListItemCell( { children, item: clientId, onLayout } ) {
 	const { blocksLayouts, updateBlocksLayouts } = useBlockListContext();
+	const { rootClientId } = useSelect(
+		( select ) => {
+			const { getBlockRootClientId } = select( blockEditorStore );
+			return { rootClientId: getBlockRootClientId( clientId ) };
+		},
+		[ clientId ]
+	);
```
React Native `VirtualizedList` expects a [React class component or render function](https://reactnative.dev/docs/0.69/virtualizedlist#cellrenderercomponent). Passing a function that returned a React Element resulted in an inability to open the inserter from a inner block appender — multiple re-renders kept resetting `isOpen`. 

In order to pass the `BlockListItemCell` component, the `rootClientId` had to be selected here instead. It does cause me concern as to whether this is will degrade performance.

I welcome thoughts on performance or possible alternative approaches.

geriux replied:
I think it'll be ok performance wise

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -346,142 +320,74 @@ export class BlockList extends Component {
-		return {
-			clearSelectedBlock,
-			insertBlock,
-			replaceBlock,
```
`replaceBlock` appeared unused and was removed.

dcalhoun commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -346,142 +320,74 @@ export class BlockList extends Component {
+					accessibilityLabel={ __( 'Add paragraph block' ) }
+					testID={ __( 'Add paragraph block' ) }
+					onPress={ () => {
+						const paragraphBlock = createBlock( 'core/paragraph' );
```
In order to satisfy React Native `VirtualizedList`'s expectation of a [React class component or render function](https://reactnative.dev/docs/0.69/virtualizedlist#listfootercomponent), the creation of the Paragraph block was moved to the event hander, rather than in the render.

dcalhoun reviewed (commented):
I generated installable [iOS](https://github.com/wordpress-mobile/WordPress-iOS/pull/20639#issuecomment-1535558316) and [Android](https://github.com/wordpress-mobile/WordPress-Android/pull/18369#issuecomment-1535483249) builds for testing purposes. 

There may be opportunity to optimize performance. I have not yet measured with React Profiler to see if there are differences between the previous class component.  

I attempted to create a Reassure test, but abandoned it for now after encountering some difficult errors.

dcalhoun commented:
I used the React Profiler to observe render count and duration for a few interactions. It appears to be comparable to the previous class component. 

| Action | Before (class) | After (function) |
| - | - | - |
| Type | ![type-before](https://user-images.githubusercontent.com/438664/236497688-6c06791d-1ebc-4d00-b287-cfdfb0c80147.png) | ![type-after](https://user-images.githubusercontent.com/438664/236497739-cee055d3-d0de-4e4b-999f-2f930b375c7a.png) |
| Insert | ![insert-before](https://user-images.githubusercontent.com/438664/236497776-da9988d4-4186-41fe-9eb7-f5254929d7ab.png) | ![insert-after](https://user-images.githubusercontent.com/438664/236497802-213a2a77-cec6-478a-aef6-a5189e88c0c8.png) |
| Drag | ![drag-before](https://user-images.githubusercontent.com/438664/236497841-3b3ab1f6-eb2b-4e04-8edc-de61858eda84.png) | ![drag-after](https://user-images.githubusercontent.com/438664/236497864-8e69fad1-7fd9-4dda-a800-08f358629f1f.png) |

geriux commented on packages/block-editor/src/components/block-list/index.native.js:
```diff
@@ -492,17 +398,36 @@ const EmptyListComponentCompose = compose( [
-] )( EmptyListComponent );
+	} );
+
+	if ( renderFooterAppender || renderAppender === false ) {
+		return null;
+	}
+
+	return (
+		<View style={ styles.defaultAppender }>
+			<ReadableContentView
+				align={
+					renderAppender ? WIDE_ALIGNMENTS.alignments.full : undefined
+				}
+			>
+				<BlockListAppender
+					rootClientId={ rootClientId }
+					renderAppender={ renderAppender }
+					showSeparator={ shouldShowInsertionPoint }
+				/>
+			</ReadableContentView>
+		</View>
+	);
```
Should we update the usage of `ReadableContentView` in favor of the `useEditorWrapperStyles` hook? Or do we plan to do it in a separate PR?

Something like:

```suggestion
	const align = renderAppender ? WIDE_ALIGNMENTS.alignments.full : undefined;
	const [ wrapperStyles ] = useEditorWrapperStyles( { align } );

	if ( renderFooterAppender || renderAppender === false ) {
		return null;
	}

	const containerStyles = [ styles.defaultAppender, wrapperStyles ];

	return (
		<View style={ containerStyles }>
			<BlockListAppender
				rootClientId={ rootClientId }
				renderAppender={ renderAppender }
				showSeparator={ shouldShowInsertionPoint }
			/>
		</View>
	);
```

dcalhoun replied:
> Should we update the usage of `ReadableContentView` in favor of the `useEditorWrapperStyles` hook? Or do we plan to do it in a separate PR?

My preference is to keep this PRs focused solely on converting this file from a class to a functional component. My hope is that will make code review easier and avoid complex merge conflicts if we split future work between different contributors. 

I planned for additional PRs for [subtasks](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5617#issuecomment-1532380331) and https://github.com/wordpress-mobile/gutenberg-mobile/issues/5618.

geriux replied:
Awesome! That sounds good, since I didn't see anything in the list mentioning `ReadableContentView` I thought I'd mentioned it. 👍 Thanks!

geriux reviewed (commented):
Nice work! I'm still reviewing it and I'd like to do some testing but I left a comment related to the `EmptyList` component. Let me know what you think!

dcalhoun committed:
fix: Scope timer run by `waitForStoreResolvers`

Running all timers could create testing failures in scenarios wheres
where inner blocks were inserted, e.g. Social Icons. It is not known
why, but scoping from "all" to "pending" timers resolved the following
errors, without breaking any existing tests.

Unknown origin:

```
Warning: You seem to have overlapping act() calls, this is not supported. Be sure to await previous act() calls before making a new one.
```

Originating from `waitForStoreResolvers`:

```
Ran 100000 timers, and there are still more! Assuming we've hit an infinite recursion and bailing out...
```

Originating from `packages/block-editor/src/components/block-list/test/index.native.js`:

```
thrown: "Exceeded timeout of 5000 ms for a test.
```

dcalhoun committed:
test: BlockList renders footer appender

dcalhoun committed:
typo: Add code comment

Added to align with sibling tests.

dcalhoun committed:
test: Expand empty BlockList appender tests

geriux commented on test/native/integration-test-helpers/wait-for-store-resolvers.js:
```diff
@@ -25,7 +25,7 @@ export async function waitForStoreResolvers( fn ) {
 
 		// Advance all timers allowing store resolvers to resolve.
-		act( () => jest.runAllTimers() );
+		act( () => jest.runOnlyPendingTimers() );
```
Out of curiosity why is this change needed?

dcalhoun replied:
I do not fully know. 😬 

For future reference, I often capture nuanced details in individual commit messages. I try to remember to surface these in pull request descriptions and comments, but I often fail to remember that step. Sorry about that. 

For this change, I captured the "why" in the body of https://github.com/WordPress/gutenberg/pull/50293/commits/34d6d78ddb17889fb69c3438eab8b425b834847c, which I will paste below for convenience. 

It would appear that inner blocks — definitely Social Links, but likely others — result in a scenario where there are _a lot_ of set timers. To my understanding, [running "pending" timers](https://jestjs.io/docs/jest-object#jestrunonlypendingtimers) means Jest runs only the _currently pending_ timers already queued at a given moment, but it will not run additional timers that are queued as a result of running that first queue of timers, i.e. chained timers. I'm unsure where these chained timers occur. Scoping the timers to "pending" resolved the test timeout errors. 

My guess is that rendering each inner block — which Social Links render several to start — generates additional resolvers or other timers. 

Do you think it is worth pursuing a root cause further? Is it worth leaving an inline comment? 

---

Running all timers could create testing failures in scenarios wheres
where inner blocks were inserted, e.g. Social Icons. It is not known
why, but scoping from "all" to "pending" timers resolved the following
errors, without breaking any existing tests.

Unknown origin:

```
Warning: You seem to have overlapping act() calls, this is not supported. Be sure to await previous act() calls before making a new one.
```

Originating from `waitForStoreResolvers`:

```
Ran 100000 timers, and there are still more! Assuming we've hit an infinite recursion and bailing out...
```

Originating from `packages/block-editor/src/components/block-list/test/index.native.js`:

```
thrown: "Exceeded timeout of 5000 ms for a test.
```

geriux replied:
> For future reference, I often capture nuanced details in individual commit messages. I try to remember to surface these in pull request descriptions and comments

I will remember that thanks! I do think is still worth mentioning something in the description and maybe referencing to the commit message.

> It would appear that inner blocks — definitely Social Links, but likely others — result in a scenario where there are a lot of set timers. To my understanding, [running "pending" timers](https://jestjs.io/docs/jest-object#jestrunonlypendingtimers) means Jest runs only the currently pending timers already queued at a given moment, but it will not run additional timers that are queued as a result of running that first queue of timers, i.e. chained timers. I'm unsure where these chained timers occur. Scoping the timers to "pending" resolved the test timeout errors.

> My guest is that rendering each inner block — which Social Links render several to start — generates additional resolvers or other timers.

I see! I wanted to connect the dots and now I understand that is related to the updated test that now adds a Social links block which surfaced this issue.

> Do you think it is worth pursuing a root cause further? Is it worth leaving an inline comment?

I don't think so, there will be a commit reference anyways for this change and this discussion. 👍 

> Running all timers could create testing failures in scenarios wheres
where inner blocks were inserted, e.g. Social Icons. It is not known
why, but scoping from "all" to "pending" timers resolved the following
errors, without breaking any existing tests.

Oh, I have seen the `Exceeded timeout of 5000 ms for a test` message before, I'm happy this solves this one and the other ones.

This change looks good to me, thanks for explaining!

fluiddot replied:
I haven't checked further this case, but based on the findings I shared [here](https://github.com/wordpress-mobile/gutenberg-mobile/issues/5742#issuecomment-1538715579), it's likely that the infinite recursion mentioned above is caused by some component using the `debounce`  function.

geriux reviewed (approved):
These changes look good to me, nice work migrating the code from a class component into a functional one 👏 

I have used the test builds for both iOS and Android to test: Block insertion, Drag&Drop, nesting blocks, color customization, auto-scroll, and VoiceOver/Talkback. All working well 🚀 

**NOTE:** This branch doesn't include the latest changes from `trunk` so some visual regression tests are failing over [this PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/5716) it'd be nice to update this branch and rerun the full tests before merging just in case.

dcalhoun committed:
Merge branch 'trunk' of github.com:WordPress/gutenberg into refactor/block-list-uses-hooks
